src/CodeGen_LLVM.cpp

```diff
@@ -602,6 +618,113 @@ void CodeGen_LLVM::compile_buffer(const Buffer &buf) {
     sym_push(buf.name() + ".buffer", global_ptr);
 }
 
+namespace {
+
+// Convert a Type into an integer key suitable for use with std::map
+int key_for_type(Type t) {
+    internal_assert(t.bits < 256 && t.width == 1);
+    return t.code * 256 + t.bits;
+}
+
+template<typename T>
+T scalar_from_constant_expr_or_die(Expr e) {
+    T v;
+    internal_assert(scalar_from_constant_expr(e, &v)) << "scalar_from_constant_expr fails for Expr " << e << "\n";
+    return v;
+}
+
+template<typename T>
+llvm::Constant *get_constant(llvm::Type *ty, Expr e) {
+    return std::numeric_limits<T>::is_integer ?
+        ConstantInt::get(ty, scalar_from_constant_expr_or_die<T>(e)) :
+        ConstantFP::get(ty, scalar_from_constant_expr_or_die<T>(e));
+}
+
+}  // namespace
+
+Constant* CodeGen_LLVM::embed_constant_expr(Expr e) {
+    if (!e.defined() || e.type().code == Type::Handle) {
+        // Handle is always emitted as "undefined"
+        return Constant::getNullValue(scalar_value_t_type->getPointerTo());
+    }
+
+    // This function is typically called exactly once (at most) per
+    // instance of CodeGen_LLVM, so it's not really worth trying to hoist
+    // out of this function. (It also uses the llvm::Type member variables
+    // from CodeGen_LLVM; keeping it here makes it easier to just re-use those.)
+    std::map<int, std::function<llvm::Constant *(Expr e)>> expr_to_llvm_constant_func_map{
```

Comment:
Hmm, you could be right (though we need a lot more else-if's than your example above). I'll experiment and see.